### ISSUE TITLE
using a type from an unanalyzed file to be a mypy error

### ISSUE THREAD

**Issue by dgoldstein0** [python/mypy#3205]
example code, in py2.7

```
import MyType from unanalyzed

def foo(bar):
  # type: (MyType) -> None
  ...
```

I would expect this to be an error: if mypy doesn't know what `MyType` is, it shouldn't be a valid type.  Instead, it assumes it's `Any` which is way more permissive behavior than I expect.

**Comment by ilevkivskyi** [comment]
This is intentional. Everything non-annotated/unknown should have type ``Any`` by default. This is one of the key ideas of _gradual_ typing, otherwise annotating a large codebase would be a non-starter.

For more control on mypy behaviour, you could use flags, for example:
```
--disallow-untyped-calls
--disallow-untyped-defs
--check-untyped-defs
--disallow-subclassing-any
--warn-return-any
```

I am closing this, since this is unlikely to change. Feel free to re-open if you have additional concerns.

**Comment by dgoldstein0** [comment]
so I understand that it is expected for unchecked imports to be Any - but what I did not expect is that we could use one of them as a type without any warning.  It doesn't sound like any of those flags you are suggesting will fix this?  Can we please reopen?  (It doesn't look like I have the ability to do this myself)

anyhow I filed this task at the suggestion of @ddfisher, who thought a flag for a warning about this could be valuable

**Comment by ilevkivskyi** [comment]
> but what I did not expect is that we could use one of them as a type without any warning. It doesn't sound like any of those flags you are suggesting will fix this?

Yes, this case is not covered by current flags. And it looks like such flag is not too difficult to implement.

**Comment by ddfisher** [comment]
(assigning to myself to hold for @ilinum)

**Comment by ddfisher** [comment]
As some additional background, this happens when the `--follow-imports=skip` and `--ignore-missing-imports` flags are set.

To implement this, I'd start by adding an additional member to [the `AnyType` class](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/types.py#L272) which tracks whether or not it's from a silent import.  I'd then grep for `ignore_missing_imports` and `follow_imports` to find where ignored module imports are turned into `Any`s, and set the member I'd added to `AnyType` there.
Next, I'd add the new `--disallow-implicit-any-types` flag.  It needs to be added to [`process_options` in `main.py`](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/main.py#L171) and to [`Options` in `options.py`](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/options.py#L16).  It should be a per-module option.
For the warning itself, I'd add checks to [`visit_func_def`](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/checker.py#L508), [`visit_overloaded_func_def`](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/checker.py#L293) and [`visit_class_def`](https://github.com/python/mypy/blob/3cd62e1795f4aa5c5d2b5de125d3244e74feb3c3/mypy/checker.py#L1072) in `checker.py`.
Finally, I'd add several tests for this functionality to the end of [`check-flags.test`](https://github.com/python/mypy/blob/master/test-data/unit/check-flags.test).
